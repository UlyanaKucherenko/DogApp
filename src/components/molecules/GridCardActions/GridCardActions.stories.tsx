import type { Meta } from '@storybook/react';
import { Box } from '@mui/material';

import { GridCardActions, GridCardActionsProps } from '.';

const meta: Meta<GridCardActionsProps> = {
  title: 'Molecules/GridCardActions',
  component: GridCardActions
};
export default meta;

export const Favorite = () => (
  <Box
    sx={{
      width: '250px',
      height: '200px',
      position: 'relative',
      backgroundColor: 'gray'
    }}
  >
    <GridCardActions id="1" isFavorite={false} cardType="favorite" />
  </Box>
);

export const Upload = () => (
  <Box
    sx={{
      width: '250px',
      height: '200px',
      position: 'relative',
      backgroundColor: 'gray'
    }}
  >
    <GridCardActions id="1" isFavorite={false} cardType="upload" />
  </Box>
);
