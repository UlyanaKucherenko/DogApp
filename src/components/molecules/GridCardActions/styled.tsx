import { Box, CardActions } from '@mui/material';
import { alpha, styled } from '@mui/material/styles';

const WrapActions = styled(CardActions)(({ theme: { palette } }) => ({
  display: 'flex',
  flexDirection: 'column',
  justifyContent: 'space-between',
  padding: '24px 24px 20px 24px',
  position: 'absolute',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  width: '100%',
  height: '100%',
  backgroundColor: 'transparent',
  transition: 'background-color 0.3s ease-in-out',
  '&:hover': {
    backgroundColor: alpha(palette.common.black, 0.45)
  }
}));

const CardActionsTop = styled(Box)(() => ({
  display: 'flex',
  justifyContent: 'flex-end',
  width: '100%'
}));

const CardActionsBottom = styled(Box)(() => ({
  width: '100%'
}));

export { WrapActions, CardActionsTop, CardActionsBottom };
