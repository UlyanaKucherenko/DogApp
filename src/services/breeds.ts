import { api } from './api';

export const breedsApi = api.injectEndpoints({
  endpoints: build => ({
    getBreeds: build.query<Breed[], void>({
      query: () => 'breeds'
    }),
    getBreedById: build.query<BreedDetails, string>({
      query: breedId => `breeds/${breedId}`
    })
  })
});

export const { useGetBreedsQuery, useGetBreedByIdQuery } = breedsApi;

export interface Breed {
  id: string;
  name: string;
  temperament: string;
}

export interface BreedDetails extends Breed {
  weight: {
    metric: string;
  };
  height: {
    metric: string;
  };
  lifeSpan: string;
  origin: string;
}
