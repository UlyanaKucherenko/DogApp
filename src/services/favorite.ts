import { api } from './api';

export interface Favorite {
  image_id: string;
  sub_id?: string;
}

export type FavoritesResponse = {
  id: number;
  user_id: string;
  image_id: string;
  sub_id: string;
  created_at: string;
  image: {
    id: string;
    url: string;
  };
};

export const favoritesApi = api.injectEndpoints({
  endpoints: build => ({
    addFavorite: build.mutation<unknown, Favorite>({
      query: body => ({
        url: `favourites`,
        method: 'POST',
        body
      }),
      invalidatesTags: [{ type: 'Favorites', id: 'LIST' }]
    }),

    deleteFavorite: build.mutation<unknown, { favourite_id: number }>({
      query: ({ favourite_id }) => ({
        url: `favourites/${favourite_id}`,
        method: 'DELETE'
      }),
      invalidatesTags: [{ type: 'Favorites', id: 'LIST' }]
    }),

    getFavorites: build.query<FavoritesResponse[], void>({
      query: () => ({
        url: `favourites`
      }),
      providesTags: (result = []) => [
        ...result.map(({ id }) => ({ type: 'Favorites', id }) as const),
        { type: 'Favorites' as const, id: 'LIST' }
      ]
    })
  })
});

export const {
  useAddFavoriteMutation,
  useDeleteFavoriteMutation,
  useGetFavoritesQuery
} = favoritesApi;
