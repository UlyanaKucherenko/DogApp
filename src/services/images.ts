import { api } from './api';

export interface Images {
  id: string;
  url: string;
  width: number;
  height: number;
  mime_type: string;
  breeds: [
    {
      id: number;
      name: string;
      weight: string;
      height: string;
      life_span: string;
      reference_image_id: string;
    }
  ];
  categories: [];
}

export interface UploadImage {
  breeds: [];
  id: string;
  url: string;
  width: number;
  height: number;
  sub_id: string;
  created_at: string;
  original_filename: string;
  breed_ids: string;
}

export type ImagesUploadResponse = {
  id: string;
  url: string;
  width: number;
  height: number;
  original_filename: string;
  pending: number;
  approved: number;
};

export interface BreedImage {
  id: string;
  url: string;
  reference_image_id: string;
}

type ImagesResponse = {
  [breedId: string]: BreedImage[];
};

export const imagesApi = api.injectEndpoints({
  endpoints: build => ({
    getBreedImages: build.query<ImagesResponse, string>({
      query: imageId => ({
        url: `images/${imageId}`
      }),
      providesTags: id => [{ type: 'Images', id }]
    }),

    getImagesId: build.query<
      ImagesResponse,
      {
        limit?: number;
        page?: number;
        mime_type?: string;
        order?: string;
        breed_id?: number;
      }
    >({
      query: ({
        limit = 10,
        page = 0,
        mime_type = 'all',
        breed_id,
        order = 'RANDOM'
      }) => ({
        url: `images/search?limit=${limit}&page=${page}&mime_types=${mime_type}&order=${order}&breed_id=${breed_id}`
      }),
      providesTags: (result = []) => [
        ...result.map(({ id }) => ({ type: 'Images', id }) as const),
        { type: 'Images' as const, id: 'LIST' }
      ]
    }),

    getImages: build.query<
      ImagesResponse,
      {
        limit?: number;
        page?: number;
        mime_type?: string;
        order?: string;
        breed_id?: number;
      }
    >({
      query: ({
        limit = 10,
        page = 0,
        mime_type = 'all',
        breed_id,
        order = 'RANDOM'
      }) => ({
        url: `images/search?limit=${limit}&page=${page}&mime_types=${mime_type}&order=${order}`
      }),
      providesTags: (result = []) => [
        ...result.map(({ id }) => ({ type: 'Images', id }) as const),
        { type: 'Images' as const, id: 'LIST' }
      ]
    }),

    uploadImage: build.mutation<Images, FormData>({
      query(formData) {
        return {
          url: `images/upload`,
          method: 'POST',
          body: formData
        };
      },
      invalidatesTags: [{ type: 'Images', id: 'LIST' }]
    }),
    uploadImagesList: build.query<
      UploadImage[],
      {
        limit?: number;
        page?: number;
      }
    >({
      query: ({ limit = 10, page = 0 }) => ({
        url: `images/?limit=${limit}&page=${page}`
      }),
      providesTags: (result = []) => [
        ...result.map(({ id }) => ({ type: 'Images', id }) as const),
        { type: 'Images' as const, id: 'LIST' }
      ]
    })
  })
});

export const {
  useGetBreedImagesQuery,
  useGetImagesIdQuery,
  useGetImagesQuery,
  useUploadImageMutation,
  useUploadImagesListQuery
} = imagesApi;
