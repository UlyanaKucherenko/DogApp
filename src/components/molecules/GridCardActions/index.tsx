import { Box, Button, IconButton } from '@mui/material';
import { useState } from 'react';

import { AddFavoriteIcon } from '../../atoms/AddFavoriteIcon';
import ModalDialog from '../../atoms/ModalDialog';
import {
  useAddFavoriteMutation,
  useDeleteFavoriteMutation
} from '../../../services/favorite';
import { WrapActions, CardActionsTop, CardActionsBottom } from './styled';

export interface GridCardActionsProps {
  id: string | number;
  isFavorite?: boolean;
  cardType?: 'favorite' | 'upload';
}

export const GridCardActions = ({
  id,
  isFavorite = false,
  cardType
}: GridCardActionsProps) => {
  const [isHovered, setIsHovered] = useState(false);
  const [isOpenDeleteModal, setIsOpenDeleteModal] = useState(false);

  const [addFavorite] = useAddFavoriteMutation();
  const [deleteFavorite] = useDeleteFavoriteMutation();

  const addToFavorite = () => {
    if (typeof id === 'string') {
      addFavorite({ image_id: id });
    }
  };

  const deleteFromFavorite = () => {
    if (typeof id === 'number') {
      deleteFavorite({ favourite_id: id });
    }
    setIsOpenDeleteModal(false);
  };
  // eslint-disable-next-line no-nested-ternary
  const iconState = isFavorite ? 'active' : isHovered ? 'hover' : 'default';

  return (
    <>
      <ModalDialog
        title="Delete the photo?"
        description="You will not be able to  recover it"
        textBtnClose="Cancel"
        textBtnConfirm="Delete"
        open={isOpenDeleteModal}
        confirm={deleteFromFavorite}
        onClose={() => setIsOpenDeleteModal(false)}
      />
      <WrapActions
        onMouseEnter={() => setIsHovered(true)}
        onMouseLeave={() => setIsHovered(false)}
      >
        <CardActionsTop>
          {isFavorite ? (
            <Box aria-label="add to favorites">
              <AddFavoriteIcon state={iconState} />
            </Box>
          ) : (
            <IconButton onClick={addToFavorite} aria-label="add to favorites">
              <AddFavoriteIcon state={iconState} />
            </IconButton>
          )}
        </CardActionsTop>
        {isHovered && (
          <CardActionsBottom>
            {cardType === 'favorite' && (
              <Box
                sx={{
                  display: 'flex',
                  justifyContent: 'flex-start'
                }}
              >
                <Button
                  variant="text"
                  color="primary"
                  onClick={() => {
                    setIsOpenDeleteModal(true);
                  }}
                >
                  delete
                </Button>
              </Box>
            )}
            {cardType === 'upload' && (
              <Box
                sx={{
                  display: 'flex',
                  justifyContent: 'flex-start'
                }}
              >
                <Button
                  variant="text"
                  color="primary"
                  onClick={() => {
                    console.log('delete item', id);
                  }}
                >
                  delete
                </Button>
              </Box>
            )}
          </CardActionsBottom>
        )}
      </WrapActions>
    </>
  );
};
