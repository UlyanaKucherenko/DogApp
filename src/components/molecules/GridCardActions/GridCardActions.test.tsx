import { composeStories } from '@storybook/react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';

import * as stories from './GridCardActions.stories';
import { store } from '../../../services/store';

describe('GridCardActions', () => {
  const { Favorite } = composeStories(stories);
  it('should match snapshot GridCardActions for FavoriteCard', () => {
    const { container } = render(
      <Provider store={store}>
        <Favorite />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
