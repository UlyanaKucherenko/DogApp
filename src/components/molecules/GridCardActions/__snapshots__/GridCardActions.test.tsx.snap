// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridCardActions should match snapshot GridCardActions for FavoriteCard 1`] = `
.emotion-0 {
  width: 250px;
  height: 200px;
  position: relative;
  background-color: gray;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  padding: 24px 24px 20px 24px;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  width: 100%;
  height: 100%;
  background-color: transparent;
  -webkit-transition: background-color 0.3s ease-in-out;
  transition: background-color 0.3s ease-in-out;
}

.emotion-1>:not(:first-of-type) {
  margin-left: 8px;
}

.emotion-1:hover {
  background-color: rgba(0, 0, 0, 0.45);
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  width: 100%;
}

.emotion-3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-3::-moz-focus-inner {
  border-style: none;
}

.emotion-3.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-3 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-3:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-3:hover {
    background-color: transparent;
  }
}

.emotion-3.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-4 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div>
  <div
    class="MuiBox-root emotion-0"
  >
    <div
      class="MuiCardActions-root MuiCardActions-spacing emotion-1"
    >
      <div
        class="MuiBox-root emotion-2"
      >
        <button
          aria-label="add to favorites"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium emotion-3"
          tabindex="0"
          type="button"
        >
          <test-file-stub
            aria-hidden="true"
            classname="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            focusable="false"
          />
          <span
            class="MuiTouchRipple-root emotion-4"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
